documentation/pdb_support_guide.markdown

```diff
@@ -0,0 +1,260 @@
+
+## Support and Troubleshooting for PuppetDB
+
+This is a technical guide for troubleshooting PuppetDB (PDB) and understanding
+its internals.
+
+## PDB Architectural Overview
+Data stored in PDB flows through four components: the terminus, a message
+queue, the actual application, and the database. The ordering is terminus ->
+application -> queue -> application -> database. It may be useful to consider
+the terminus, application plus queue, and database as residing on three
+separate machines.
+
+### Terminus
+The terminus resides on the Puppet master and redirects agent data to
+PDB in the form of "commands". PDB has four commands, as described in the
+[commands documentation][commands].
+
+### Queue
+When the terminus sends a command to PDB, the command is dumped to an ActiveMQ
+message queue on disk unprocessed, where it waits for the application to take
+it and submit it to the database. Commands may be processed out of order,
+though an approximate ordering by submission time can be loosely assumed.
+
+### Command processing
+PDB processes the queue concurrently with the number of threads specified [in
+configuration][threads]. Commands are processed differently depending on their
+type:
+
+* `store-report`: store-report commands are relatively inexpensive, consisting
+  mainly of inserting a row in the reports table.
+* `replace-catalog`: When a replace-catalog command is received, PDB will first
+  check if a more recent catalog already exists in the database for the node.
+  If so, the catalog is discarded and no action is taken. If not, PDB will
+  execute a diff of the catalog in hand and the catalog in the database, and
+  insert only the resources and edges that have changed.
+* `replace-facts`: PDB stores facts as key-value associations between "paths"
+  and "values". The term "path" refers to a specific route from the root of a
+  tree (e.g structured fact) to a leaf value, and "value" refers to the leaf
+  value itself. Conceptually, every fact is stored as a tree. To
+  illustrate, the fact
+
+      "foo" => "bar"
+
+  is stored as
+
+      "foo" => "bar"
+
+  while the fact
+
+      "foo" => {"a" => "bar", "b" => "baz"}
+
+  is stored as
+
+      "foo#~a" => "bar"
+      "foo#~b" => "baz"
+
+  For the array case, the fact
+
+      "foo" => ["bar", "baz"]
+
+  is stored as
+
+      "foo#~0" => "bar"
+      "foo#~1" => "baz"
+
+  The same rules apply recursively for larger structures. When a replace-facts
+  command is received, PDB will compare the fact in hand against the paths
+  and values in the database, add whatever new paths/values are required, and
+  delete any invalidated pairs.
+
+* `deactivate-node`: The deactivate-node command just updates a column in the
+  certnames table, and isn't likely to be the source of performance issues.
+
+### Database
+
+PDB uses PostgreSQL. The only good way to get familiar with the schema is
+to examine an [erd diagram][erd] and investigate for yourself on a running
```

Comment:
@nfagerlund this is a link to a .png. Is there anything special I need to do to make it look alright on the site?